postgres/datadog_checks/postgres/connections.py

```diff
@@ -51,12 +69,28 @@ def __init__(self, connect_fn):
                 )
         self.connect_fn = connect_fn
 
-    def get_connection(self, dbname, ttl_ms):
+    def get_connection(self, dbname: str, ttl_ms: int, timeout: int = None) -> psycopg2.extensions.connection:
+        """
+        Grab a connection from the pool if the database is already connected.
+        If max_conns is specified, and the database isn't already connected,
+        make a new connection if the max_conn limit hasn't been reached.
+        Blocks until a connection can be added to the pool,
+        and optionally takes a timeout in seconds.
+        """
+        start = datetime.datetime.now()
         self.prune_connections()
         with self._mu:
-            conn = self._conns.pop(dbname, ConnectionWithTTL(None, None))
+            conn = self._conns.pop(dbname, ConnectionInfo(None, None, None, None, None))
             db = conn.connection
             if db is None or db.closed:
+                if self.max_conns is not None and len(self._conns) == self.max_conns:
+                    # try to free space until we succeed
+                    while len(self._conns) >= self.max_conns:
+                        self.prune_connections()
+                        self.evict_lru()
+                        if timeout is not None and (datetime.datetime.now() - start).total_seconds() > timeout:
+                            raise TimeoutError
```

Comment:
This doesn't seem like the right kind of exception to raise here: https://docs.python.org/3/library/exceptions.html#TimeoutError -- referring to `timed out at the system level`

Might I suggest a custom exception defined in this module?

Reply:
I added a ConnectionPoolFullError and will raise that instead!